API test [failure](https://jenkins.ceph.com/job/ceph-api/117796/), first time seeing this
```
Python Unit-test Failure

FAIL: test_pool_list (tasks.mgr.dashboard.test_pool.PoolTest.test_pool_list)

2026-06-05 07:38:16,717.717 INFO:__main__:----------------------------------------------------------------------
2026-06-05 07:38:16,717.717 INFO:__main__:Traceback (most recent call last):
2026-06-05 07:38:16,717.717 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-api/qa/tasks/mgr/dashboard/test_pool.py", line 182, in test_pool_list
2026-06-05 07:38:16,717.717 INFO:__main__:    self.assertNotIn('pg_status', pool)
2026-06-05 07:38:16,717.717 INFO:__main__:AssertionError: 'pg_status' unexpectedly found in {'pool': 1, 'pool_name': 'rbd', 'flags': 1048577, 'flags_names': 'hashpspool,split_reads', 'type': 'replicated', 'size': 3, 'min_size': 1, 'crush_rule': 'replicated_rule', 'peering_crush_bucket_count': 0, 'peering_crush_bucket_target': 0, 'peering_crush_bucket_barrier': 0, 'peering_crush_bucket_mandatory_member': 2147483647, 'is_stretch_pool': False, 'object_hash': 2, 'pg_autoscale_mode': 'on', 'pg_num': 1, 'pg_placement_num': 1, 'pg_placement_num_target': 1, 'pg_num_target': 1, 'pg_num_pending': 1, 'last_pg_merge_meta': {'ready_epoch': 0, 'last_epoch_started': 0, 'last_epoch_clean': 0, 'source_pgid': '0.0', 'source_version': "0'0", 'target_version': "0'0"}, 'auid': 0, 'snap_mode': 'selfmanaged', 'snap_seq': 0, 'snap_epoch': 0, 'pool_snaps': [], 'quota_max_bytes': 0, 'quota_max_objects': 0, 'tiers': [], 'tier_of': -1, 'read_tier': -1, 'write_tier': -1, 'cache_mode': 'none', 'target_max_bytes': 0, 'target_max_objects': 0, 'cache_target_dirty_ratio_micro': 400000, 'cache_target_dirty_high_ratio_micro': 600000, 'cache_target_full_ratio_micro': 800000, 'cache_min_flush_age': 0, 'cache_min_evict_age': 0, 'erasure_code_profile': '', 'ec_data_shard_count': 0, 'ec_coding_shard_count': 0, 'hit_set_params': {'type': 'none'}, 'hit_set_period': 0, 'hit_set_count': 0, 'use_gmt_hitset': True, 'min_read_recency_for_promote': 0, 'min_write_recency_for_promote': 0, 'hit_set_grade_decay_rate': 0, 'hit_set_search_last_n': 0, 'grade_table': [], 'stripe_width': 0, 'expected_num_objects': 0, 'fast_read': False, 'options': {}, 'application_metadata': ['rbd'], 'read_balance': {'score_type': 'Fair distribution', 'score_acting': 4.0, 'score_stable': 4.0, 'optimal_score': 0.75, 'raw_score_acting': 3.0, 'raw_score_stable': 3.0, 'primary_affinity_weighted': 1.0, 'average_primary_affinity': 1.0, 'average_primary_affinity_weighted': 1.0}, 'create_time': '2026-06-05T06:51:35.516774+0000', 'last_change': '11', 'last_force_op_resend': '0', 'last_force_op_resend_prenautilus': '0', 'last_force_op_resend_preluminous': '0', 'removed_snaps': '[]', 'nonprimary_shards': '{}', 'pg_status': {'active+clean': 1}, 'stats': {'stored': {'latest': 0, 'rate': 0.0, 'rates': []}, 'stored_data': {'latest': 0, 'rate': 0.0, 'rates': []}, 'stored_omap': {'latest': 0, 'rate': 0.0, 'rates': []}, 'objects': {'latest': 0, 'rate': 0.0, 'rates': []}, 'kb_used': {'latest': 0, 'rate': 0.0, 'rates': []}, 'bytes_used': {'latest': 0, 'rate': 0.0, 'rates': []}, 'data_bytes_used': {'latest': 0, 'rate': 0.0, 'rates': []}, 'omap_bytes_used': {'latest': 0, 'rate': 0.0, 'rates': []}, 'percent_used': {'latest': 0.0, 'rate': 0.0, 'rates': []}, 'max_avail': {'latest': 141709557760, 'rate': 0.0, 'rates': []}, 'quota_objects': {'latest': 0, 'rate': 0.0, 'rates': []}, 'quota_bytes': {'latest': 0, 'rate': 0.0, 'rates': []}, 'dirty': {'latest': 0, 'rate': 0.0, 'rates': []}, 'rd': {'latest': 0, 'rate': 0.0, 'rates': []}, 'rd_bytes': {'latest': 0, 'rate': 0.0, 'rates': []}, 'wr': {'latest': 0, 'rate': 0.0, 'rates': []}, 'wr_bytes': {'latest': 0, 'rate': 0.0, 'rates': []}, 'compress_bytes_used': {'latest': 0, 'rate': 0.0, 'rates': []}, 'compress_under_bytes': {'latest': 0, 'rate': 0.0, 'rates': []}, 'stored_raw': {'latest': 0, 'rate': 0.0, 'rates': []}, 'avail_raw': {'latest': 425128686592, 'rate': 0.0, 'rates': []}}}
2026-06-05 07:38:16,717.717 INFO:__main__:
```

--- a/src/pybind/mgr/dashboard/frontend/cypress/e2e/page-helper.po.ts
+++ b/src/pybind/mgr/dashboard/frontend/cypress/e2e/page-helper.po.ts
@@ -290,8 +290,13 @@ export abstract class PageHelper {
 
   filterTable(name: string, option: string) {
     this.waitDataTableToLoad();
-    cy.get('select#filter_name').select(name);
-    cy.get('select#filter_option').select(option);
+    cy.get('[data-testid=filter-button]').click();
+    cy.get('cds-popover-content')
+      .should('be.visible')
+      .within(() => {
+        cy.get(`[data-testid="filter-select-${name}"]`).find('select').select(option);
+        cy.get('[data-testid="apply-filters"]').click();
+      });
   }
 
   setPageSize(size: string) {

--- a/src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/datatable.module.ts
+++ b/src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/datatable.module.ts
@@ -19,7 +19,9 @@ import {
   InputModule,
   GridModule,
   LayoutModule,
-  InlineLoadingModule
+  InlineLoadingModule,
+  PopoverModule,
+  TooltipModule
 } from 'carbon-components-angular';
 import AddIcon from '@carbon/icons/es/add/16';
 import FilterIcon from '@carbon/icons/es/filter/16';
@@ -109,7 +111,9 @@ import { TableDetailDirective } from './directives/table-detail.directive';
     InputModule,
     GridModule,
     LayoutModule,
-    InlineLoadingModule
+    InlineLoadingModule,
+    PopoverModule,
+    TooltipModule
   ],
   declarations: [
     TableComponent,

--- a/src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.html
+++ b/src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.html
@@ -51,31 +51,82 @@
       </cds-table-toolbar-search>
       <!-- end search -->
       <!-- column filters -->
-      <ng-container *ngIf="columnFilters.length !== 0">
-        <div class="filter-wrapper">
-          <svg [cdsIcon]="icons.filter"
-               [size]="icons.size16"
-               class="cds--toolbar-action__icon"></svg>
-          <cds-select (valueChange)="onSelectFilter($event)"
-                      display="inline"
-                      id="filter_name">
-            <ng-container *ngFor="let filter of columnFilters">
-              <option [value]="filter.column.name"
-                      [selected]="filter.column.name === selectedFilter.column.name">{{ filter.column.name }}</option>
-            </ng-container>
-          </cds-select>
-          <cds-select (valueChange)="onChangeFilter($event)"
-                      display="inline"
-                      id="filter_option">
-            <option *ngIf="!selectedFilter.value"
-                    i18n>Any</option>
-            <ng-container *ngFor="let option of selectedFilter.options">
-              <option [value]="option.raw"
-                      [selected]="option.raw === selectedFilter?.value?.raw">{{ option.formatted }}</option>
-            </ng-container>
-          </cds-select>
+       @if (columnFilters.length !== 0) {
+        <div cdsPopover
+             [isOpen]="openFilterPopover"
+             [autoAlign]="true"
+             [caret]="false">
+          <cds-icon-button
+            kind="ghost"
+            class="toolbar-action"
+            (click)="toggleFilterPopover()"
+            data-testid="filter-button">
+            <svg [cdsIcon]="icons.filter"
+                 [size]="icons.size16"
+                 class="cds--toolbar-action__icon"></svg>
+          </cds-icon-button>
+          <cds-popover-content [cdsLayer]="0">
+            <div
+              cdsGrid
+              [fullWidth]="true"
+              [condensed]="true"
+              [narrow]="true">
+            @for (filter of columnFilters; track filter.column.name; let i = $index) {
+            <div
+              cdsRow
+              [cdsLayer]="2"
+              class="cds-mt-4">
+
+              <!-- filter names column -->
+              <div
+                cdsCol
+                [columnNumbers]="{sm: 4, md: 4}">
+                <cds-text-label>Filter
+                  <input cdsText
+                         readonly
+                         size="sm"
+                         [id]="'filter_key_' + i"
+                         [name]="'filter_key_' + i"
+                         [value]="filter.column.name">
+                </cds-text-label>
+              </div>
+
+              <!-- filter values column -->
+              <div
+                cdsCol
+                [columnNumbers]="{sm: 4, md: 4}">
+                <cds-select (valueChange)="onChangeFilter($event, filter)"
+                            [id]="'filter_value_' + i"
+                            label="Value"
+                            size="sm"
+                            [attr.data-testid]="'filter-select-' + filter.column.name"
+                            i18n-label>
+                  <option i18n>Any</option>
+                  @for (option of filter.options; track option.raw) {
+                    <option [value]="option.raw"
+                            [selected]="option.raw === selectedFilters[filter.column.name]">
+                      {{ option.formatted }}
+                    </option>
+                  }
+
+                </cds-select>
+              </div>
+            </div>
+          }
+          <div cdsRow
+               class="cds-mt-4 cds-mb-4">
+
+            <button cdsButton="primary"
+                    size="sm"
+                    data-testid="apply-filters"
+                    (click)="onSubmitFilter()">
+              <ng-container i18n>Apply filters</ng-container>
+            </button>
+          </div>
+          </div>
+          </cds-popover-content>
         </div>
-      </ng-container>
+      }
       <!-- end column filters -->
       <!-- refresh button -->
       <cds-icon-button
@@ -126,27 +177,51 @@
     </cds-table-toolbar-content>
   </cds-table-toolbar>
   <!-- filter chips for column filters -->
-  <div class="d-flex justify-content-end align-items-center filter-tags"
-       *ngIf="toolHeader && columnFiltered">
-  <div class="d-flex gap-2">
-    <ng-container *ngFor="let filter of columnFilters">
-      <cds-tag *ngIf="filter.value"
-               type="outline"
-               class="align-self-center">
-        <span class="me-2">{{ filter.column.name }}: {{ filter.value.formatted }}</span>
-        <button class="cds--tag__close-icon"
-                (click)="onChangeFilter(filter)">
-          <svg [cdsIcon]="icons.destroy"
-               [size]="icons.size16"></svg>
+   @if (toolHeader && columnFiltered) {
+    <div class="filter-tags">
+      <div
+        class="align-items-center"
+        cdsStack="horizontal"
+        [gap]="2">
+        @for (filter of activeFilters | slice:0:3; track filter.column.name) {
+          @if (filter.value) {
+            <cds-tag-filter
+              closeButtonLabel="Clear"
+              i18n-closeButtonLabel
+              (close)="onRemoveFilter(filter)">
+              {{ filter.column.name }}={{ filter.value.formatted }}
+            </cds-tag-filter>
+          }
+        }
+        @if (activeFilters.length > 3) {
+          <cds-tooltip [description]="remainingFiltersTpl">
+            <cds-tag>
+              +{{ activeFilters.length - 3 }}
+            </cds-tag>
+
+            <ng-template #remainingFiltersTpl>
+              <div
+                cdsStack="vertical"
+                [gap]="2">
+              @for (filter of activeFilters | slice:3; track filter.column.name) {
+                <cds-tag-filter
+                  closeButtonLabel="Clear"
+                  i18n-closeButtonLabel
+                  (close)="onRemoveFilter(filter)">
+                  {{ filter.column.name }}={{ filter.value.formatted }}
+                </cds-tag-filter>
+              }
+              </div>
+            </ng-template>
+          </cds-tooltip>
+        }
+        <button cdsButton="ghost"
+                (click)="onClearFilters($event)">
+          <ng-container i18n>Clear filters</ng-container>
         </button>
-      </cds-tag>
-    </ng-container>
-    <button cdsButton="ghost"
-            (click)="onClearFilters($event)">
-      <ng-container i18n>Clear filters</ng-container>
-    </button>
-  </div>
-  </div>
+      </div>
+    </div>
+  }
   <!-- end filter chips for column filters -->
   <table cdsTable
          [sortable]="sortable"

--- a/src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.scss
+++ b/src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.scss
@@ -64,8 +64,8 @@
 }
 
 .filter-tags {
-  background-color: var(--cds-layer-accent);
-  border-bottom: 1px solid var(--cds-layer-active);
+  background-color: var(--cds-layer-01);
+  border-top: var(--cds-spacing-01) solid var(--cds-layer-02);
   color: var(--cds-text-primary);
 }
 

--- a/src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.spec.ts
+++ b/src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.spec.ts
@@ -134,8 +134,8 @@ describe('TableComponent', () => {
     ) => {
       component.search = search;
       _.forEach(changes, (change) => {
-        component.onSelectFilter(change.filter.column.name);
-        component.onChangeFilter(change.value || undefined);
+        component.onChangeFilter(change.value || undefined, change.filter);
+        component.onSubmitFilter();
       });
       expect(component.rows).toEqual(results);
       component.onClearSearch();

--- a/src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.ts
+++ b/src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.ts
@@ -25,7 +25,12 @@ import { CellTemplate } from '~/app/shared/enum/cell-template.enum';
 import { Icons, IconSize, EMPTY_STATE_IMAGE } from '~/app/shared/enum/icons.enum';
 
 import { CdTableColumn } from '~/app/shared/models/cd-table-column';
-import { CdTableColumnFilter } from '~/app/shared/models/cd-table-column-filter';
+import {
+  CdTableColumnFilter,
+  CdTableColumnFilterOption,
+  CdTableColumnSelectedFilter,
+  CdTableColumnStagedFilter
+} from '~/app/shared/models/cd-table-column-filter';
 import { CdTableColumnFiltersChange } from '~/app/shared/models/cd-table-column-filters-change';
 import { CdTableFetchDataContext } from '~/app/shared/models/cd-table-fetch-data-context';
 import { CdTableSelection } from '~/app/shared/models/cd-table-selection';
@@ -434,6 +439,14 @@ export class TableComponent implements AfterViewInit, OnInit, OnChanges, OnDestr
   editStates: EditState = {};
   formGroup: CdFormGroup = new CdFormGroup({});
 
+  openFilterPopover = false;
+  stagedFilters: CdTableColumnStagedFilter = {};
+  selectedFilters: CdTableColumnSelectedFilter = {};
+
+  get activeFilters() {
+    return this.columnFilters.filter((filter) => filter.value);
+  }
+
   constructor(
     // private ngZone: NgZone,
     private cdRef: ChangeDetectorRef,
@@ -754,6 +767,10 @@ export class TableComponent implements AfterViewInit, OnInit, OnChanges, OnDestr
     this.tableColumns = this.localColumns;
   }
 
+  toggleFilterPopover() {
+    this.openFilterPopover = !this.openFilterPopover;
+  }
+
   initColumnFilters() {
     let filterableColumns = _.filter(this.localColumns, { filterable: true });
     filterableColumns = [...filterableColumns, ...this.extraFilterableColumns];
@@ -767,6 +784,13 @@ export class TableComponent implements AfterViewInit, OnInit, OnChanges, OnDestr
       };
     });
     this.selectedFilter = _.first(this.columnFilters);
+    this.initSelectedColumnFilters();
+  }
+
+  private initSelectedColumnFilters() {
+    this.columnFilters.forEach((filter) => {
+      this.selectedFilters[filter.column.name] = filter.value?.raw;
+    });
   }
 
   private createColumnFilterOption(
@@ -805,14 +829,43 @@ export class TableComponent implements AfterViewInit, OnInit, OnChanges, OnDestr
     });
   }
 
-  onSelectFilter(filter: string) {
-    const value = this.columnFilters.find((x) => x.column.name === filter);
-    this.selectedFilter = value;
+  // saving the filters to a staged variable so they are not applied immediately
+  onChangeFilter(selectedValue: string, filter: CdTableColumnFilter) {
+    const filterName = filter.column.name;
+    const selectedFilter = this.selectedFilters[filterName];
+    const newSelectedFilter = selectedFilter === selectedValue ? undefined : selectedValue;
+
+    this.selectedFilters[filterName] = newSelectedFilter;
+
+    const option = filter.options.find(
+      (x: CdTableColumnFilterOption) => x.raw === newSelectedFilter
+    );
+    this.stagedFilters[filterName] = option;
   }
 
-  onChangeFilter(filter: string) {
-    const option = this.selectedFilter.options.find((x) => x.raw === filter);
-    this.selectedFilter.value = _.isEqual(this.selectedFilter.value, option) ? undefined : option;
+  onSubmitFilter() {
+    this.columnFilters.forEach((filter) => {
+      const filterName = filter.column.name;
+
+      if (this.stagedFilters.hasOwnProperty(filterName)) {
+        filter.value = this.stagedFilters[filterName];
+        this.selectedFilter = filter;
+      }
+    });
+
+    this.stagedFilters = {};
+    this.updateFilter();
+    this.openFilterPopover = false;
+  }
+
+  onRemoveFilter(filter: CdTableColumnFilter) {
+    const filterName = filter.column.name;
+    filter.value = undefined;
+    this.selectedFilters[filterName] = undefined;
+    delete this.stagedFilters[filterName];
+    if (this.selectedFilter?.column.name === filterName) {
+      this.selectedFilter = undefined;
+    }
     this.updateFilter();
   }
 
@@ -1288,6 +1341,7 @@ export class TableComponent implements AfterViewInit, OnInit, OnChanges, OnDestr
     });
     this.selectedFilter = _.first(this.columnFilters);
     this.updateFilter();
+    this.initSelectedColumnFilters();
   }
 
   updateFilter() {

--- a/src/pybind/mgr/dashboard/frontend/src/app/shared/models/cd-table-column-filter.ts
+++ b/src/pybind/mgr/dashboard/frontend/src/app/shared/models/cd-table-column-filter.ts
@@ -2,6 +2,19 @@ import { CdTableColumn } from './cd-table-column';
 
 export interface CdTableColumnFilter {
   column: CdTableColumn;
-  options: { raw: string; formatted: string }[]; // possible options of a filter
-  value?: { raw: string; formatted: string }; // selected option
+  options: CdTableColumnFilterOption[]; // possible options of a filter
+  value?: CdTableColumnFilterOption; // selected option
+}
+
+export interface CdTableColumnStagedFilter {
+  [filterName: string]: CdTableColumnFilterOption;
+}
+
+export interface CdTableColumnSelectedFilter {
+  [filterName: string]: string | undefined;
+}
+
+export interface CdTableColumnFilterOption {
+  raw: string;
+  formatted: string;
 }

